pkg/kncloudevents/message_sender_test.go

```diff
@@ -194,6 +196,345 @@ func TestHTTPMessageSenderSendWithRetriesWithSingleRequestTimeout(t *testing.T)
+		//
+
+		{
+			name: "small max 429 without Retry-After",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:     http.StatusTooManyRequests,
+			responseFormat: None,
+			wantReqCount:   3,
+		},
+		{
+			name: "small max 429 with Retry-After seconds",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:      http.StatusTooManyRequests,
+			responseFormat:  Seconds,
+			responseSeconds: 4,
+			wantReqCount:    3,
+		},
+		{
+			name: "small max 429 with Retry-After date",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:      http.StatusTooManyRequests,
+			responseFormat:  Date,
+			responseSeconds: 2,
+			wantReqCount:    3,
+		},
+		{
+			name: "small max 429 with invalid Retry-After",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:     http.StatusTooManyRequests,
+			responseFormat: Invalid,
+			wantReqCount:   3,
+		},
+		{
+			name: "small max 503 with Retry-After seconds",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:      http.StatusServiceUnavailable,
+			responseFormat:  Seconds,
+			responseSeconds: 4,
+			wantReqCount:    3,
+		},
+		{
+			name: "small max 500 without Retry-After",
+			config: &RetryConfig{
+				RetryMax:              2,
+				RetryAfterMaxDuration: &smallRetryAfterMaxDuration,
+			},
+			statusCode:     http.StatusInternalServerError,
+			responseFormat: None,
+			wantReqCount:   3,
+		},
+	}
+
+	// Execute The TestCases
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+
+			// Consistent CheckRetry & Backoff Implementation For All TestCases
+			paddingDuration := 250 * time.Millisecond // Test Execution Allowance For Resending Requests
+			backoffDuration := 500 * time.Millisecond // Constant Backoff Time For Simplified Testing
+			tc.config.CheckRetry = SelectiveRetry
+			tc.config.Backoff = func(attemptNum int, resp *http.Response) time.Duration { return backoffDuration }
+
+			// Determine The Response Retry-After Backoff Duration From TestCase
+			var responseDuration time.Duration
+			switch tc.responseFormat {
+			case None, Invalid:
+			case Seconds, Date:
+				if tc.responseSeconds > 0 {
+					responseDuration = time.Duration(tc.responseSeconds) * time.Second
+				}
+			default:
+				assert.Fail(t, "TestCase with unsupported ResponseFormat '%v'", tc.responseFormat)
+			}
+
+			// Tracking Variables
+			var previousReqTime time.Time
+			var reqCount int32
+
+			// Create A Test HTTP Server Capable Of Validating Request Interim Durations
+			server := httptest.NewServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
```

Comment:
This test is a beast, I hope you're around to maintain it 😄

Reply:
+1, let's make the code below more modular and simpler to reason about.

Reply:
Yeah, agreed - was trying to be thorough with the testing.  I added a comment at top and I will maintain (adjust timings or rewrite/delete/etc as needed in the future ; )

Reply:
above comment was in response to ben (commenting race condition ; ).   I'll try to restructure it a bit but inherently it's a complicated test (custom server config and server and client validation).